File: ProgrammingChallenges#Book/Chapter 1/Graphical Editor/solutions/GraphicalEditor.py
from sys import stdin
import queue

class Position:
    def __init__(self, x, y):
        self.x = x
        self.y = y


class GEditor:
    def __init__(self):
        self.M = 0
        self.N = 0
        self.picture = [['' for x in range(255)] for y in range(255)]
        self.marked  = [[False for x in range(255)] for y in range(255)]

    def processCommand(self, command):
        def I():
            # commandI(int M, int N)
            self.commandI(int(command[1]), int(command[2]))

        def C():
            self.commandC()

        def L():
            # commandL(int X, int Y, char C)
            self.commandL(int(command[1])-1, int(command[2])-1, command[3])

        def V():
            # commandV(int X, int Y1, int Y2, char C)
            self.commandV(int(command[1])-1, int(command[2])-1, int(command[3])-1, command[4])

        def H():
            # commandH(int X1, int X2, int Y, char C)
            self.commandH(int(command[1])-1, int(command[2])-1, int(command[3])-1, command[4])

        def H():
            # commandH(int X1, int X2, int Y, char C)
            self.commandH(int(command[1])-1, int(command[2])-1, int(command[3])-1, command[4])

        def K():
            # commandK(int X1, int X2, int Y1, int Y2, char C)
            self.commandK(int(command[1])-1, int(command[2])-1, int(command[3])-1, int(command[4])-1, command[5]);

        def F():
            # commandF(int X, int Y, char C)
            self.commandF(int(command[1])-1, int(command[2])-1, command[3]);

        def S():
            # commandS(string Name)
            self.commandS(command[1]);
            
        switch = {
            'I': I,
            'C': C,
            'L': L,
            'V': V,
            'H': H,
            'K': K,
            'F': F,
            'S': S
        }

        if (command[0] in switch.keys()):
            switch[command[0]]()


    def commandI(self, M, N):
        self.M = M
        self.N = N

        for i in range(0, self.N):
            for j in range(0, self.M):
                self.picture[i][j] = 'O'


    def commandC(self):
        for i in range(0, self.N):
            for j in range(0, self.M):
                self.picture[i][j] = 'O'


    def commandL(self, X, Y, C):
        self.picture[Y][X] = C


    def commandV(self, X, Y1, Y2, C):
        if (Y1 > Y2):
            Y1 = Y1+Y2
            Y2 = Y1-Y2
            Y1 = Y1-Y2
        
        for i in range(Y1, Y2+1):
            self.picture[i][X] = C

    def commandH(self, X1, X2, Y, C):
        if (X1 > X2):
            X1 = X1+X2
            X2 = X1-X2
            X1 = X1-X2

        for i in range(X1, X2+1):
            self.picture[Y][i] = C

    def commandK(self, X1, Y1, X2, Y2, C):
        if (X1 > X2):
            X1 = X1+X2
            X2 = X1-X2
            X1 = X1-X2

        if (Y1 > Y2):
            Y1 = Y1+Y2
            Y2 = Y1-Y2
            Y1 = Y1-Y2

        for i in range(Y1, Y2+1):
            for j in range(X1, X2+1):
                self.picture[i][j] = C


    def commandF(self, X, Y, C):
        self.resetMarkedValues()

        pixel = self.picture[Y][X]

        self.picture[Y][X] = C
        self.marked[Y][X] = True

        pixelQueue = queue.Queue()
        pixelQueue.put(Position(X, Y))

        while (not pixelQueue.empty()):
            temp = pixelQueue.get()
            
            if (temp.y-1 >= 0 and not self.marked[temp.y-1][temp.x] and self.picture[temp.y-1][temp.x] == pixel):
                self.marked[temp.y-1][temp.x] = True
                self.picture[temp.y-1][temp.x] = C
                pixelQueue.put(Position(temp.x,temp.y-1))
            if (temp.y+1 < self.N and not self.marked[temp.y+1][temp.x] and self.picture[temp.y+1][temp.x] == pixel):
                self.marked[temp.y+1][temp.x] = True
                self.picture[temp.y+1][temp.x] = C
                pixelQueue.put(Position(temp.x,temp.y+1))
            if (temp.x-1 >= 0 and not self.marked[temp.y][temp.x-1] and self.picture[temp.y][temp.x-1] == pixel):
                self.marked[temp.y][temp.x-1] = True
                self.picture[temp.y][temp.x-1] = C
                pixelQueue.put(Position(temp.x-1,temp.y))
            if (temp.x+1 < self.M and not self.marked[temp.y][temp.x+1] and self.picture[temp.y][temp.x+1] == pixel):
                self.marked[temp.y][temp.x+1] = True
                self.picture[temp.y][temp.x+1] = C
                pixelQueue.put(Position(temp.x+1,temp.y))


    def resetMarkedValues(self):
        for i in range(0, self.N):
            for j in range(0, self.M):
                self.marked[i][j] = False
    

    # Writes the picture in the file Name.
    def commandS(self, Name):
        print(Name)
        self.printGEditor();


    def printGEditor(self):
        for i in range(0, self.N):
            for j in range(0, self.M):
                print(self.picture[i][j], end = "")
            print()



# Main program
def main():    
    geditor = GEditor()

    for line in stdin:
        command = line.split()

        if (command[0] == 'X'):
            break

        geditor.processCommand(command)

main()

File: ProgrammingChallenges#Book/Chapter 1/Graphical Editor/solutions/GraphicalEditor.cpp
#include <stdio.h>
#include <string.h>
#include <queue>

using namespace std;

class Position {
    public:
        int x, y;

        Position(int x, int y) {
            this->x = x;
            this->y = y;
        }
};

class GEditor {
    public:
        int M, N;
        char picture[250][250];
        int marked[250][250];

        GEditor() {}

        void processCommand(char command) {
            char C;
            char Name[100];
            int M, N, X, Y, X1, X2, Y1, Y2;


            switch (command) {
                case 'I':
                    // commandI(int M, int N)
                    scanf("%d %d", &M, &N);
                    this->commandI(M, N);
                    break;
                case 'C':
                    this->commandC();
                    break;
                case 'L':
                    // commandI(int X, int Y, char C)
                    scanf("%d %d %c", &X, &Y, &C);
                    this->commandL(X-1, Y-1, C);
                    break;
                case 'V':
                    // commandV(int X, int Y1, int Y2, char C)
                    scanf("%d %d %d %c", &X, &Y1, &Y2, &C);
                    this->commandV(X-1, Y1-1, Y2-1, C);
                    break;
                case 'H':
                    // commandH(int X1, int X2, int Y, char C)
                    scanf("%d %d %d %c", &X1, &X2, &Y, &C);
                    this->commandH(X1-1, X2-1, Y-1, C);
                    break;
                case 'K':
                    // commandK(int X1, int X2, int Y1, int Y2, char C)
                    scanf("%d %d %d %d %c", &X1, &X2, &Y1, &Y2, &C);
                    this->commandK(X1-1, X2-1, Y1-1, Y2-1, C);
                    break;
                case 'F':
                    // commandF(int X, int Y, char C)
                    scanf("%d %d %c", &X, &Y, &C);
                    this->commandF(X-1, Y-1, C);
                    break;
                case 'S':
                    // commandS(string Name)
                    scanf("%s", Name);
                    this->commandS(Name);
                    break;
            }
        }

        void commandI(int M, int N) {
            this->M = M;
            this->N = N;

            for (int i = 0; i < this->N; i++) {
                for (int j = 0; j < this->M; j++) {
                    this->picture[i][j] = 'O';
                }
            }
        }


        void commandC() {
            for (int i = 0; i < this->N; i++) {
                for (int j = 0; j < this->M; j++) {
                    this->picture[i][j] = 'O';
                }
            }
        }

        void commandL(int X, int Y, char C) {
            this->picture[Y][X] = C;
        }

        void commandV(int X, int Y1, int Y2, char C){
            if (Y1 > Y2) {
                Y1 = Y1+Y2;
                Y2 = Y1-Y2;
                Y1 = Y1-Y2;
            }
            for (int i = Y1; i <= Y2; i++)
                this->picture[i][X] = C;
        }

        void commandH(int X1,int X2,int Y, char C){
            if (X1 > X2) {
                X1 = X1+X2;
                X2 = X1-X2;
                X1 = X1-X2;
            }
            for (int i = X1; i <= X2; i++)
                this->picture[Y][i] = C;
        }

        void commandK(int X1, int Y1, int X2, int Y2, char C){
            if (X1 > X2){
                X1 = X1+X2;
                X2 = X1-X2;
                X1 = X1-X2;
            }
            if (Y1 > Y2) {
                Y1 = Y1+Y2;
                Y2 = Y1-Y2;
                Y1 = Y1-Y2;
            }
            for (int i = Y1; i <= Y2; i++) {
                for (int j = X1; j <= X2; j++) {
                    this->picture[i][j] = C;
                }
            }
        }

        void commandF(int X, int Y, char C){
            this->resetMarkedValues();

            char pixel = this->picture[Y][X];

            this->picture[Y][X] = C;
            this->marked[Y][X] = true;
            queue<Position*> pixelQueue;
            pixelQueue.push(new Position(X, Y));
            while(!pixelQueue.empty()){
                Position* temp = pixelQueue.front();
                pixelQueue.pop();
                if(temp->y-1 >= 0 && !this->marked[temp->y-1][temp->x] && this->picture[temp->y-1][temp->x] == pixel){
                    this->marked[temp->y-1][temp->x]=true;
                    this->picture[temp->y-1][temp->x]=C;
                    pixelQueue.push(new Position(temp->x,temp->y-1));
                }
                if(temp->y+1 < this->N && !this->marked[temp->y+1][temp->x] && this->picture[temp->y+1][temp->x] == pixel){
                    this->marked[temp->y+1][temp->x]=true;
                    this->picture[temp->y+1][temp->x]=C;
                    pixelQueue.push(new Position(temp->x,temp->y+1));
                }
                if(temp->x-1 >= 0 && !this->marked[temp->y][temp->x-1] && this->picture[temp->y][temp->x-1] == pixel){
                    this->marked[temp->y][temp->x-1]=true;
                    this->picture[temp->y][temp->x-1]=C;
                    pixelQueue.push(new Position(temp->x-1,temp->y));
                }
                if(temp->x+1 < this->M && !this->marked[temp->y][temp->x+1] && this->picture[temp->y][temp->x+1] == pixel){
                    this->marked[temp->y][temp->x+1] = true;
                    this->picture[temp->y][temp->x+1] = C;
                    pixelQueue.push(new Position(temp->x+1,temp->y));
                }
            }
        }

        void resetMarkedValues() {
            for (int i = 0; i < this->N; i++) {
                for (int j = 0; j < this->M; j++) {
                    this->marked[i][j] = false;
                }
            }
        }

        /* Writes the picture in the file Name. */
        void commandS(char Name[100]) {
            printf("%s\n", Name);
            this->printGEditor();
        }

        void printGEditor(){
            for (int i = 0; i < this->N; i++) {
                for (int j = 0; j < this->M; j++) {
                    printf("%c", this->picture[i][j]);
                }
                printf("\n");
            }
        }
};

/* Main program */
int main() {
    int i, j, n, m, resultIndex = 1;

    char inputLine[30];
    char command;

    GEditor* geditor =  new GEditor();

    while (scanf("%c", &command) != EOF) {
        if (command == 'X') {
            break;
        }

        geditor->processCommand(command);
    }

    return 0;
}